drivers/pwm/pwm_nrfx.c

```diff
@@ -283,64 +282,53 @@ static int pwm_nrfx_init(const struct device *dev)
 
 		seq_values_ptr_get(dev)[i] = PWM_NRFX_CH_VALUE(0, inverted);
 	}
-
-	nrfx_err_t result = nrfx_pwm_init(&config->pwm,
-					  &config->initial_config,
-					  NULL,
-					  NULL);
-	if (result != NRFX_SUCCESS) {
-		LOG_ERR("Failed to initialize device: %s", dev->name);
-		return -EBUSY;
-	}
-
-	return 0;
 }
 
-#ifdef CONFIG_PM_DEVICE
-static void pwm_nrfx_uninit(const struct device *dev)
+static void pwm_suspend(const struct device *dev)
 {
 	const struct pwm_nrfx_config *config = dev->config;
 
-	nrfx_pwm_uninit(&config->pwm);
+	nrfx_pwm_stop(&config->pwm, false);
+	while (!nrfx_pwm_stopped_check(&config->pwm)) {
+	}
 
 	memset(dev->data, 0, sizeof(struct pwm_nrfx_data));
+	(void)pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
 }
 
 static int pwm_nrfx_pm_action(const struct device *dev,
 			      enum pm_device_action action)
 {
-	const struct pwm_nrfx_config *config = dev->config;
-	int ret = 0;
+	if (action == PM_DEVICE_ACTION_RESUME) {
+		pwm_resume(dev);
+	} else if (IS_ENABLED(CONFIG_PM_DEVICE) && (action == PM_DEVICE_ACTION_SUSPEND)) {
+		pwm_suspend(dev);
+	} else {
+		return -ENOTSUP;
+	}
 
-	switch (action) {
-	case PM_DEVICE_ACTION_RESUME:
-		ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
-		if (ret < 0) {
-			return ret;
-		}
-		ret = pwm_nrfx_init(dev);
-		break;
+	return 0;
+}
 
-	case PM_DEVICE_ACTION_SUSPEND:
-		pwm_nrfx_uninit(dev);
+static int pwm_nrfx_init(const struct device *dev)
+{
+	const struct pwm_nrfx_config *config = dev->config;
+	nrfx_err_t err;
 
-		ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
-		if (ret < 0) {
-			return ret;
-		}
-		break;
+	ANOMALY_109_EGU_IRQ_CONNECT(NRFX_PWM_NRF52_ANOMALY_109_EGU_INSTANCE);
 
-	default:
-		return -ENOTSUP;
+	if (IS_ENABLED(CONFIG_PM_DEVICE)) {
```

Comment:
One last thing, this should rather be `CONFIG_PM_DEVICE_RUNTIME`, because only then the device is expected to be initialized as suspended (when the runtime PM is auto-enabled for it) but the SUSPEND action is not called, right?